phoenix-core/src/main/java/org/apache/phoenix/iterate/ClientHashAggregatingResultIterator.java

```diff
@@ -0,0 +1,189 @@
+        Tuple tuple = wrapKeyValueAsResult(KeyValueUtil.newKeyValue(key, SINGLE_COLUMN_FAMILY, SINGLE_COLUMN, AGG_TIMESTAMP, value, 0, value.length));
+        return tuple;
+    }
+
+    @Override
+        public void close() throws SQLException {
+        keyIterator = null;
+        keyList = null;
+        hash = null;
+        memoryChunk.resize(0);
+        resultIterator.close();
+    }
+
+    @Override
+        public Aggregator[] aggregate(Tuple result) {
+        Aggregator[] rowAggregators = aggregators.getAggregators();
+        aggregators.reset(rowAggregators);
+        aggregators.aggregate(rowAggregators, result);
+        return rowAggregators;
+    }
+
+    @Override
+        public void explain(List<String> planSteps) {
+        resultIterator.explain(planSteps);
+    }
+
+    @Override
+        public String toString() {
+        return "ClientHashAggregatingResultIterator [resultIterator="
+            + resultIterator + ", aggregators=" + aggregators + ", groupByExpressions="
+            + groupByExpressions + "]";
+    }
+
+    // Copied from ClientGroupedAggregatingResultIterator
+    protected ImmutableBytesWritable getGroupingKey(Tuple tuple, ImmutableBytesWritable ptr) throws SQLException {
+        try {
+            ImmutableBytesWritable key = TupleUtil.getConcatenatedValue(tuple, groupByExpressions);
+            ptr.set(key.get(), key.getOffset(), key.getLength());
+            return ptr;
+        } catch (IOException e) {
+            throw new SQLException(e);
+        }
+    }
+
+    // Copied from ClientGroupedAggregatingResultIterator
+    protected Tuple wrapKeyValueAsResult(KeyValue keyValue) {
+        return new MultiKeyValueTuple(Collections.<Cell> singletonList(keyValue));
+    }
+
+    private HashMap<ImmutableBytesWritable, Aggregator[]> populateHash() throws SQLException {
+
+        hash = new HashMap<ImmutableBytesWritable, Aggregator[]>(HASH_AGG_INIT_SIZE, 0.75f);
+        final int aggSize = aggregators.getEstimatedByteSize();
+
+        for (Tuple result = resultIterator.next(); result != null; result = resultIterator.next()) {
+            ImmutableBytesWritable key = new ImmutableBytesWritable(UNITIALIZED_KEY_BUFFER);
+            key = getGroupingKey(result, key);
+            Aggregator[] rowAggregators = hash.get(key);
+            if (rowAggregators == null) {
+                long hashSize = SizedUtil.sizeOfMap(hash.size() + 1, key.getSize(), aggSize);
+                if (hashSize > memoryChunk.getSize() + CLIENT_HASH_AGG_MEMORY_CHUNK_SIZE) {
+                    // This will throw InsufficientMemoryException if necessary
+                    memoryChunk.resize(hashSize + CLIENT_HASH_AGG_MEMORY_CHUNK_SIZE);
+                }
+
+                // Abort if we exceed memory threshold/2
+                // We use threshold/2 to leave room for the subsequent sort
+                if (thresholdBytes > 0) {
+                    if (hashSize > thresholdBytes/2) {
```

Comment:
I don't think we need this since we're tracking memory in a different way now. This would put a constraint on each client side aggregation (rather than memory usage as a whole). Do you think that additional constraint is necessary and if so why? FYI, you can also set an absolute byte amount for memory usage (but this is mainly used in tests). If you want to keep this, then let's create a different config parameter in QueryServices and a default in QueryServiesOptions. The spooling threshold is not a memory constraint. It's tells Phoenix when to start spooling to disk during a client sort, so it's not really appropriate to use here.

Reply:
Done, removed.